### ISSUE TITLE
How can I used sincnet for speaker verification task?

### ISSUE THREAD

**Issue by bharatsubedi41** [mravanelli/SincNet#29]
Thank you for your contribution.
In paper you are mentioning about the speaker verification performance but in  code I did not found any code related to speaker verification. Please can you explain me how can i implement verification.

**Comment by mravanelli** [comment]
Hi!
thank you very much for your interest in my research. We recently release
the code to compute the d-vectors (see the repository) that is the most
important component of the speaker verification system. We are moving all
the speaker recognition part (i.e., speaker id + speaker verification)
within the pytorch-kaldi project (
https://github.com/mravanelli/pytorch-kaldi). We plan to put there the full
experimental pipeline.

Mirco

<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
Virus-free.
www.avast.com
<https://www.avast.com/sig-email?utm_medium=email&utm_source=link&utm_campaign=sig-email&utm_content=webmail>
<#DAB4FAD8-2DD7-40BB-A1B8-4E2AA1F9FDF2>

**Comment by bharatsubedi41** [comment]
Thank you for your  response.
I am planning to implement speaker verification in keras. Is it available in keras or not?

**Comment by mravanelli** [comment]
As far as I remember someone has implemented Sincnet in Keras.

**Comment by Aki58** [comment]
Hi,
In paper, there are two approaches for speaker verification using trained speaker id to compute d vectors and another is dnn-class. there is a line "DNN-class models perform significantly better than d-vectors". Since my model is trained for speaker id task now I want to use it for speaker verification task. 
So i need to know the approaches in both directions

1. using compute dvector i can find the d vector of new audio files of speaker so how to use the compute d vector for verification whether the new audio corresponds to a given speaker as i want to verify the new test audio. In simple way a bit explanation on how to use the dvectors for speaker verification.

2. Explain the DNN-class models.

Thanks,
Akash

**Comment by mravanelli** [comment]
Hi,
the DNN-class model works well but it is not really scalable. The idea is
to retrain a novel speaker-id network every time a new speaker comes. To do
speaker verification, we just consider the average probability of this
speaker as speaker verification score. The difference with d-vector is that
d-vector is computed and then "frozen", while DNN_class requires a new
training for every new speaker. In a real application, d-vector is the
preferred solution.

Mirco

**Comment by Aki58** [comment]
how to use the compute d vector for speaker verification? I have a sincnet model trained for n speakers. now we have new speaker audio for verification. then what is the flow to verify whether that speaker is specified speaker or not. In the verification paper it is given to find cosine simalrity between both the d vectors so one d vector is of the new speaker file and where i can find another d vector. what is the flow for speaker verification?

**Comment by Aki58** [comment]
"During the evaluation phase, we first extract the normalized dvector from the test utterance. Then we compute the cosine distance between the test d-vector and the claimed speaker’s d-vector". the test d-vector can be obtained using the test audio file and compute_d_vector.py , now to find the the claimed speaker’s d-vector. we need to compute speaker specific d-vector. how can we save speaker specific d-vector?

**Comment by mravanelli** [comment]
The speaker specific d-vector can be computed summing up all the d-vectors
for all the sentences uttered by that speaker.

**Comment by Aki58** [comment]
If there is a new speaker that Speaker Identification model has not seen then to enroll him in speaker verification system we need to take dvector for all the sentences uttered by that speaker using the previously trained speaker Id model and take average of it? And next time when the speaker comes then we will calculate the dvector of the sentence uttered at that time and calculate cosine similarity with the previously stored d vector to verify the speaker?
